src/java/org/apache/cassandra/service/reads/repair/BlockingPartitionRepair.java

```diff
@@ -0,0 +1,248 @@
+import org.apache.cassandra.locator.ReplicaCollection;
+import org.apache.cassandra.locator.ReplicaList;
+import org.apache.cassandra.locator.ReplicaSet;
+import org.apache.cassandra.locator.Replicas;
+import org.apache.cassandra.metrics.ReadRepairMetrics;
+import org.apache.cassandra.net.IAsyncCallback;
+import org.apache.cassandra.net.MessageIn;
+import org.apache.cassandra.net.MessageOut;
+import org.apache.cassandra.net.MessagingService;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.tracing.Tracing;
+
+public class BlockingPartitionRepair extends AbstractFuture<Object> implements IAsyncCallback<Object>
+{
+    private final Keyspace keyspace;
+    private final DecoratedKey key;
+    private final ConsistencyLevel consistency;
+    private final ReplicaList participants;
+    private final Map<Replica, Mutation> pendingRepairs;
+    private final CountDownLatch latch;
+
+    private volatile long mutationsSentTime;
+
+    public BlockingPartitionRepair(Keyspace keyspace, DecoratedKey key, ConsistencyLevel consistency, Map<Replica, Mutation> repairs, int maxBlockFor, ReplicaList participants)
+    {
+        this.keyspace = keyspace;
+        this.key = key;
+        this.consistency = consistency;
+        this.pendingRepairs = new ConcurrentHashMap<>(repairs);
+        this.participants = participants;
+
+        // here we remove empty repair mutations from the block for total, since
+        // we're not sending them mutations
+        int blockFor = maxBlockFor;
+        for (Replica participant: participants)
+        {
+            // remote dcs can sometimes get involved in dc-local reads. We want to repair
+            // them if they do, but they shouldn't interfere with blocking the client read.
+            if (!repairs.containsKey(participant) && shouldBlockOn(participant.getEndpoint()))
+                blockFor--;
+        }
+
+        // there are some cases where logically identical data can return different digests
+        // For read repair, this would result in ReadRepairHandler being called with a map of
+        // empty mutations. If we'd also speculated on either of the read stages, the number
+        // of empty mutations would be greater than blockFor, causing the latch ctor to throw
+        // an illegal argument exception due to a negative start value. So here we clamp it 0
+        latch = new CountDownLatch(Math.max(blockFor, 0));
+    }
+
+    @VisibleForTesting
+    long waitingOn()
+    {
+        return latch.getCount();
+    }
+
+    @VisibleForTesting
+    boolean isLocal(InetAddressAndPort endpoint)
+    {
+        return ConsistencyLevel.isLocal(endpoint);
+    }
+
+    private boolean shouldBlockOn(InetAddressAndPort endpoint)
+    {
+        return !consistency.isDatacenterLocal() || isLocal(endpoint);
+    }
+
+    @VisibleForTesting
+    void ack(InetAddressAndPort from)
+    {
+        if (shouldBlockOn(from))
+        {
+            // TODO from is NOT Replica!
```

Comment:
Starting to look like this is still in progress. Definitely not going to work to remove by the wrong type key.